src/PointwiseFunctions/Xcts/LongitudinalOperator.cpp

```diff
@@ -0,0 +1,89 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "PointwiseFunctions/Elasticity/PotentialEnergy.hpp"
+
+#include <algorithm>
+#include <cstddef>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+
+namespace Xcts {
+
+/// \cond
+template <typename DataType>
+void longitudinal_operator(const gsl::not_null<tnsr::II<DataType, 3>*> result,
+                           const tnsr::ii<DataType, 3>& strain,
+                           const tnsr::II<DataType, 3>& inv_metric) noexcept {
+  for (size_t i = 0; i < 3; ++i) {
+    for (size_t j = 0; j <= i; ++j) {
+      // Unroll first iteration of the loop over `k` to avoid filling the result
+      // with zero initially
+      result->get(i, j) = 2. *
+                          (inv_metric.get(i, 0) * inv_metric.get(j, 0) -
+                           inv_metric.get(i, j) * get<0, 0>(inv_metric) / 3.) *
+                          get<0, 0>(strain);
+      for (size_t l = 1; l < 3; ++l) {
+        result->get(i, j) +=
+            2. *
+            (inv_metric.get(i, 0) * inv_metric.get(j, l) -
+             inv_metric.get(i, j) * inv_metric.get(0, l) / 3.) *
+            strain.get(0, l);
+      }
+      for (size_t k = 1; k < 3; ++k) {
+        for (size_t l = 0; l < 3; ++l) {
+          const auto projection =
+              2. * (inv_metric.get(i, k) * inv_metric.get(j, l) -
+                    inv_metric.get(i, j) * inv_metric.get(k, l) / 3.);
+          result->get(i, j) += projection * strain.get(k, l);
+        }
+      }
+    }
+  }
+}
+
+template <typename DataType>
+void longitudinal_operator_flat_cartesian(
+    const gsl::not_null<tnsr::II<DataType, 3>*> result,
+    const tnsr::ii<DataType, 3>& strain) noexcept {
+  // Compute trace term in 0-0 component of the result
+  get<0, 0>(*result) = get<0, 0>(strain);
+  for (size_t d = 1; d < 3; ++d) {
+    get<0, 0>(*result) += strain.get(d, d);
+  }
+  get<0, 0>(*result) *= -2. / 3.;
+  // Copy trace term to other diagonal components and complete with non-trace
+  // term
+  for (size_t d = 1; d < 3; ++d) {
+    result->get(d, d) = get<0, 0>(*result) + 2. * strain.get(d, d);
+  }
```

Comment:
Compute trace into **2,2** component.  Then this loop can go from d=0 to 3 without needing a special case for (0,0) at the end.